contracts/libraries/VaultLifecycleSTETH.sol

```diff
@@ -0,0 +1,747 @@
+        address recipient,
+        uint256 withdrawAmount
+    ) internal returns (uint256 yieldTokenBalance) {
+        IERC20 collateral = IERC20(collateralToken);
+
+        yieldTokenBalance = collateral.balanceOf(address(this));
+        uint256 yieldTokensToWithdraw =
+            dsmin(yieldTokenBalance, withdrawAmount);
+        if (yieldTokensToWithdraw > 0) {
+            collateral.safeTransfer(recipient, yieldTokensToWithdraw);
+        }
+    }
+
+    /**
+     * @notice Withdraws `asset` from vault
+     * @param weth is the weth address
+     * @param asset is the vault asset address
+     * @param collateralToken is the address of the collateral token
+     * @param recipient is the recipient
+     * @param withdrawAmount is the withdraw amount in terms of yearn tokens
+     * @param yieldTokenBalance is the collateral token (yvWETH) balance of the vault
+     */
+    function withdrawBaseToken(
+        address weth,
+        address asset,
+        address collateralToken,
+        address recipient,
+        uint256 withdrawAmount,
+        uint256 yieldTokenBalance
+    ) internal {
+        uint256 underlyingTokensToWithdraw =
+            IWSTETH(collateralToken).getStETHByWstETH(
+                withdrawAmount.sub(yieldTokenBalance)
+            );
+
+        transferAsset(
+            weth,
+            asset,
+            payable(recipient),
+            underlyingTokensToWithdraw
+        );
+    }
+
+    /**
+     * @notice Unwraps the necessary amount of the yield-bearing yearn token
+     *         and transfers amount to vault
+     * @param amount is the amount of `asset` to withdraw
+     * @param asset is the vault asset address
+     * @param collateralToken is the address of the collateral token
+     */
+    function unwrapYieldToken(
+        uint256 amount,
+        address asset,
+        address collateralToken
+    ) external {
+        uint256 assetBalance = IERC20(asset).balanceOf(address(this));
+
+        uint256 amountToUnwrap =
+            IWSTETH(collateralToken).getWstETHByStETH(
+                dsmax(assetBalance, amount).sub(assetBalance)
+            );
+
+        if (amountToUnwrap > 0) {
+            IWSTETH wsteth = IWSTETH(collateralToken);
+            // Unrap to stETH
+            wsteth.unwrap(wsteth.balanceOf(address(this)));
+
+            // CRV SWAP HERE from steth -> eth
+        }
+    }
+
+    /**
+     * @notice Wraps the necessary amount of the base token to the yield-bearing yearn token
+     * @param collateralToken is the address of the collateral token
+     */
+    function wrapToYieldToken(address collateralToken) external {
+        ISTETH steth = ISTETH(IWSTETH(collateralToken).stETH());
+        if (address(this).balance > 0) {
+            // Send eth to Lido, recieve steth
+            steth.submit{value: address(this).balance}(address(this));
+            // Wrap to wstETH
+            IWSTETH(collateralToken).wrap(steth.balanceOf(address(this)));
```

Comment:
Optimization: We can use the value returned by `submit` to get the number of shares returned. This is to avoid calling `steth.balanceOf`, which consumes 2100 gas.

```sol
uint shares = steth.submit{value: ethBalance}(address(this));
IWSTETH(collateralToken).wrap(shares); // double check this
```

https://etherscan.io/address/0x20dc62d5904633cc6a5e34bec87a048e80c92e97#code

Reply:
actually, cant do this because what if somebody deposits steth directly.